proposals/migrate-over-ipfs.md

```diff
@@ -0,0 +1,152 @@
+version will need to download one of these migrations and many of Brave's users are in China.
+
+#### Assumptions &amp; hypotheses
+_What must be true for this project to matter?_
+
+- IPFS in Brave needs to matter.
+- A significant number of Brave users need to be in China.
+- Fetching migrations over IPFS needs to be reliable.
+
+#### User workflow example
+_How would a developer or user use this new capability?_
+
+- When starting the go-ipfs daemon with the `--migrate` flag, the migrations would be fetched over IPFS instead of HTTPs.
+- When starting the go-ipfs daemon in a bundled application like IPFS Desktop or Brave, the user shouldn't notice anything except that we make no connections to https://ipfs.io when starting go-ipfs the first time after upgrading.
+
+#### Impact
+_How would this directly contribute to web3 dev stack product-market fit?_
+
+- This would unblock updating go-ipfs in Brave, allowing us to ship bug fixes and new features to Brave users.
+- This would prove out IPFS as a way to ship code/updates to users.
+
+#### Leverage
+_How much would nailing this project improve our knowledge and ability to execute future projects?_
+
+This is ones step towards our goal of updating go-ipfs itself over IPFS. In terms of knowledge, not much.
+
+#### Confidence
+_How sure are we that this impact would be realized? Label from [this scale](https://medium.com/@nimay/inside-product-introduction-to-feature-priority-using-ice-impact-confidence-ease-and-gist-5180434e5b15)_.
+
+10? We're not shipping a new release to brave until we have _a_ way to ship repo migrations to users in China.
+
+## Project definition
+
+#### Brief plan of attack
+
+**Design sketch:**
+
+When migrating a repo:
+
+1. Create a new _temporary_ repo (go-ipfs can't read the current repo because it's an older version).
+    a. If the local go-ipfs node uses a swarm key, skip to step 3.a (download over HTTPs).
+2. Start a new _temporary_ go-ipfs node in the temporary repo.
+    a. This node should not listen for inbound connections as it has no way to know which ports/transports should be configured (can't read the config).
+    b. This node should not expose an API/gateway.
+3. Download the required migration binaries using the temporary go-ipfs node.
+    a. If this fails, download over HTTPs.
+4. Migrate the main go-ipfs node's repo.
+5. Start the main go-ipfs node and _transfer_ (but don't pin) the downloaded migrations into the main repo so that others can download it from this node.
+6. Shut down the temporary node and delete the temporary repo.
+
+Provide the following configuration section:
+
+```js
+{
+    "Migration": {
+        // When true, always use IPFS. When false, never use IPFS. When unset, pick the default
+        // behavior.
+        "UseIPFS": true|false,
+        // When true or unset, use the default gateway. When false, don't use a gateway. When a
+        // string, use the specified gateway.
+        "UseGateway": true|false|"my-gateway",
```

Comment:
I feel this could be simpler and more flexible, how about something like:

```suggestion
        // Control how migration should be downloaded.
        // "ipfs" means fetching over IPFS, everything else indicates the name of HTTP gateway.
        // When left empty, an implicit default of ["ipfs", "dweb.link", "ipfs.io"] is used.
        "DownloadPolicy": []
```

Reply:
Hm. Yeah, that works. I'd still like a "default-gateway" option but this is nice and flexible (and allows the user to specify the order).